### PR TITLE
Refactoring of Tokens Interface for DASB

### PR DESCRIPTION
**Pull request by Chaanks**
Work in progress ...

### CODE REVIEW

File: benchmarks/DASB/LibriSpeech/ASR-refactor-tokens/train.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+from hyperpyyaml import load_hyperpyyaml
+from pathlib import Path
+
+base_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../"))
+sys.path.append(base_dir)
+
+
+logger = logging.getLogger(__name__)
+
+_CACHE = {"size": 0}
```
Comment (lucadellalib) [inline benchmarks/DASB/LibriSpeech/ASR-refactor-tokens/train.py:28]: This is not used here

File: benchmarks/DASB/LibriSpeech/ASR-refactor/train.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+
+        # Add waveform augmentation if specified.
+        if stage == sb.Stage.TRAIN and hasattr(self.hparams, "wav_augment"):
+            wavs, wav_lens = self.hparams.wav_augment(wavs, wav_lens)  # [B, T]
+
+        # compute features
+        # Extract tokens (cache them at first epoch if augmentation is disabled)
+        key = tuple(sorted(batch.id))
```
Comment (lucadellalib) [inline benchmarks/DASB/LibriSpeech/ASR-refactor/train.py:44]: This caching code works well only if batch_size=1 or batch_size > 1 and sorting is not random.

File: benchmarks/DASB/utils/tokens.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+Authors
+---------
+* Jarod Duret, 2024
+"""
+
+import math
+import logging
+import pathlib as pl
+import kaldiio
```
Comment (lucadellalib) [inline benchmarks/DASB/utils/tokens.py:12]: What are the advantages of using `kaldiio` format over, for example a text file with list of utterance ids and a folder of `.npz` archives, one for each utterance?

**Comment by mravanelli** [comment]
@lucadellalib, please make sure the current proposal is compliant with your tasks. If not, propose or uush modifications. Also, make sure the feature storing and reading is not a bottleneck in the case of shared filesystems (e.g, those in compute canada).

File: benchmarks/DASB/run_extraction.sh [hunk 4]
```diff
@@ -0,0 +1,114 @@
+#!/bin/bash
+
+###########################################################
+# Script to extracts and save tokens from dataset.
+#
+# Usage:
+# ./ $run_extraction.sh  --data_folder LibriSpeech --output_folder results/LibriSpeech/ASR/encodec/LSTM --tokenizer encidec --dataset LibriSpeech
```
Comment (Chaanks) [inline benchmarks/DASB/run_extraction.sh]: Typo: encidec -> encodec

File: benchmarks/DASB/run_hparam_optimization.sh [hunk 5]
```diff
@@ -0,0 +1,416 @@
+#!/bin/bash
+
+###########################################################
+# Hyperparameter Tuning Script for EEG Model with Orion
```
Comment (Chaanks) [inline benchmarks/DASB/run_hparam_optimization.sh:4]: The doc need to be updated

File: benchmarks/DASB/utils/aggregate_results.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                float(metrics[cnt, :].mean()),
+                float(metrics[cnt, :].std()),
+            )
+            line = line.replace("<values>", values_line)
+            cnt = cnt + 1
+        print(line)
+
+
+if __name__ == "__main__":
+    output_folder = sys.argv[1]
```
Comment (Chaanks) [inline benchmarks/DASB/utils/aggregate_results.py:128]: Check for missing or invalid arguments

File: benchmarks/DASB/LibriSpeech/ASR-on-the-fly/hparams/LSTM/dac.yaml [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+
+# LSTM
+activation: !name:torch.nn.Sigmoid
+dnn_layers: 2
+dnn_neurons: 1024
+dropout: 0.2
+output_neurons: 31
+
+# BPE parameters
```
Comment (Chaanks) [inline benchmarks/DASB/LibriSpeech/ASR-on-the-fly/hparams/LSTM/dac.yaml:122]: Duplicate comment

**Comment by poonehmousavi** [comment]
Thanks, @Chaanks, for the PR! All recipes have been tested, and the main changes are as follows:

1. **Offline and On-the-Fly Token Extraction for LibriSpeech ASR Recipe**  
   - Added on-the-fly token extraction as a demonstration.  
   - For other recipes, we will adhere to the offline approach as it is faster.
   - Time for one epoch(libirspeech100) both  On-the-Fly and offline versions , enocdec, low botrate:
    ```
    online: __main__ - Model execution time: 1340.385495 seconds
    offline: __main__ - Model execution time: 399.695498 seconds
    ```

2. **TokenizerInterface Enhancements**  
   - Ensured the input and output of the `encode` and `decode` functions are consistent across all tokenizers.  
   - Decoupled dependencies from SpeechBrain. Any future changes in SpeechBrain will only require updates to the `tokenizer_interface` file, not individual recipes.

3. **Hyperparameter Tuning Pipeline**  
   - Integrated a TPE-based hyperparameter tuning pipeline.  
   - Enabled multiple runs and averaged results for robust evaluation.

4. **Updated README File**  
   - Improved documentation to reflect the latest changes and enhancements.

**Pull request merged**